# 자료구조와 반복문

## 함수의 그래프 작성

함수 $f(x)$의 그래프를 그리기 위해서는 다음 과정을 순서대로 수행하면 된다.

1. 함수 값을 계산할 $x$값을 결정한다. 이를 $x_1 < x_2 < \dots < x_n$이라 하자. 

2. 모든 $x_i$에 대해서 $f(x_i)$를 계산한다.

3. 그래프에 좌표가 $(x_i, f(x_i))$인 점을 찍는다.

4. 인접한 점을 직선으로 연결한다.

정밀한 그림을 작성하려면 $n$이 커야 하고, $x_i$, $i=1, 2, ..., n$ 간의 간격이 작아야 한다.

여기서는 $\theta = 2$일 때 $0 \le x \le 10$ 구간에서 함수 $f(x) = \theta e^{-\theta x}$의 그래프를 작성하기 위해 위에서 언급한 1단계와 2단계를 수행하는 프로그램을 작성해보자.
$0 \le x \le 10$에 사이에 있는 모든 $x$에 대해 함수 값을 계산할 수 없기 때문에 $0 \le x \le 10$에서 함수 값을 계산할 유한개의 $x$를 결정해야 한다.
주어진 구간을 일정한 간격으로 등분하는 값을 사용하는 것이 일반적이다.
예를 들어, $0 \le x \le 10$ 구간을 100 등분하는 101개의 값에서 함수 값을 계산할 수 있다.

이때 101개의 값을 서로 다른 변수에 저장한다면 101개의 변수 이름을 지어야 하며 이로 인해 프로그램도 상당히 어려워지고 복잡해진다.
따라서 **하나의 이름에 여러 개의 값을 저장하고 저장된 값을 편리하게 이용할 수 있게 해주는 자료형**이 필요하다.
이런 기능을 가진 자료형을 **자료구조**(data structure)라고 한다.

**자료구조는 개념적으로는 수학의 벡터(vector)나 행렬과 유사**하지만 이보다 더 많은 종류가 있으며 프로그래밍에 다양한 용도로 사용된다.
모든 프로그래밍 언어는 여러 가지 자료구조를 제공한다.
대부분의 프로그래밍 언어가 자료구조를 라이브러리로 지원하지만 **파이썬은 언어 자체에 기본적인 자료구조가 포함되어 있기 때문에 사용하기 편하고 실행속도도 빠르다.**
파이썬이 제공하는 기본적인 자료구조는 `list`, `set`, `tuple`, `dict` 등인데 각각은 개념적으로 목록, 집합, 순서쌍, 사전에 해당한다.

## 리스트의 생성과 사용

다음은 0과 1 사이를 10등분하는 11개 값을 이름이 x인 리스트에 저장하는 예이다.

**Note: 파이썬의 기본 자료구조를 만들 때 `[]`, `()`, `{}`와 같은 여러 종류의 괄호를 이용한다.
어떤 괄호를 사용하느냐에 따라 다른 자료구조가 생성되고, 자료구조가 달라지면 이용하는 방법과 기능이 달라진다는 것을 기억하기 바란다.
`[]`은 리스트, `()`은 튜플, `{}`은 집합과 사전에 사용된다.**

In [0]:
x = [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0] # BRACKET, 대괄호
print(type(x))
print(x)

<class 'list'>
[0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]


리스트에 저장된 값을 이용할 때는 몇 번째 값인가를 지정해야 하는데 이를 **인덱스**(index)라고 한다.
첫 번째 저장된 값의 인덱스는 0이고 그 다음 값의 인덱스는 1, 2, ...이다.
따라서 **리스트 자료구조는 값이 저장된 순서가 중요한 자료구조**이다.
리스트 외에도 순서가 중요한 자료구조는 순서쌍이다.
인덱스는 리스트 이름 다음에 `[]`를 쓰고 그 안에 기재한다.

In [0]:
print(x[0])
print(x[1])
print(x[10])

0
0.1
1.0


일단 리스트에 값을 저장하면 저장한 값의 개수는 기억할 필요가 없다.
리스트에 저장된 값의 개수가 필요하면 `len()` 내장함수(built-in function)로 언제든지 알아낼 수 있기 때문이다.

In [0]:
print(len(x))  
print(x[len(x)-1])  
print(x[len(x)-3])

11
1.0
0.8


**연산 결과가 정수가 되는  표현식**을 인덱스로 사용할 수도 있다.

In [0]:
i = 2
print(x[i**2 + 3])

0.7


인덱스를 이용하여 리스트에 저장된 값을 변경할 수 있다.
저장된 값을 수정할 수 있는 자료형을 **mutable**하다고 하는데 리스트는 **mutable**한 자료구조이다.
저장된 값을 수정할 수 없는 자료형을 **immutable**하다고 한다.

In [0]:
old_value = x[len(x)-1]
x[len(x)-1] = 1000000
print(x)

x[len(x)-1] = old_value
print(x)

[0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1000000]
[0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]


리스트 자료구조에 저장할 값들이 반드시 동일한 자료형일 필요는 없다.
다음은 하나의 리스트에 여러 자료형의 값을 저장하는 예이다.

In [0]:
student = ['name park', 74.5, 12345, True]
print(student)

['name park', 74.5, 12345, True]


## 슬라이싱

리스트에 저장된 값 한 개가 아니라 여러 개의 값을 추출할 수도 있다.
이때에는 추출하고자 하는 값들의 인덱스를 `[]` 속에
```python
시작_인덱스:종료_인덱스:인덱스_증감량
```
와 같은 형식으로 주면 된다.
그러면
```python
시작_인덱스, 시작_인덱스 + 인덱스_증감량, 시작_인덱스 + 인덱스_증감량 * 2, ...
```
에 해당하면서 종료_인덱스보다 작은 인덱스에 해당하는 값이 추출되어 **별도의 리스트**로 만들어 진다.
인덱스 증감량이 1이면
```python
시작_인덱스:종료_인덱스
```
이 간단히 기재할 수 있다.
만약 시작 인덱스 이후의 모든 값을 추출할 때는

```python
시작_인덱스:
```

와 같이 기재할 수 있고, 시작 인덱스가 0이면 다음과 같이 시작 인덱스를 생략할 수 있다.

```python
:종료_인덱스
```

```python
시작_인덱스:종료_인덱스:인덱스증감량
```
와 같이 기재한 것을 **슬라이스**(slice)라고 하고, 슬라이스를 이용하여 리스트의 원소 일부분을 추출하여 새로운 리스트를 만드는 것을 **슬라이싱**(slicing)이라 한다.

In [0]:
print(x)

[0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]


In [0]:
x_front = x[:5]  # x[0:5], x[0:5:1]
x_middle = x[5:7]  # x[5:7:1]
x_rear = x[7:]  # x[7:(len(x)-1):1]

x_even = x[1::2]  # x[1:(len(x)-1):2]
x_odd = x[::2]  # x[0:(len(x)-1):2]
x_full = x[:]  # x[0:(len(x)-1):1], 모두==> 리스트의 복사본을 생성 

print(x)
print(x_front)
print(x_middle)
print(x_rear)
print(x_even)
print(x_odd)

[0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
[0, 0.1, 0.2, 0.3, 0.4]
[0.5, 0.6]
[0.7, 0.8, 0.9, 1.0]
[0.1, 0.3, 0.5, 0.7, 0.9]
[0, 0.2, 0.4, 0.6, 0.8, 1.0]


## 리스트 간의 연산

리스트 간에는 `+` 연산을 할 수 있다.
두 리스트에 대해 `+` 연산을 하면 두 리스트를 연결한 긴 리스트가 **새로** 만들어진다.
**기존의 리스트는 계속 존재한다.**
`+` 연산자가 정수나 실수에 대해서는 덧셈 연산을 산술 연산자이지만 리스트에 대해서는 그렇지 않다.
프로그래밍 언어에서는 **피연산자의 자료형에 따라 연산자가 다르게 작동할 수 있다**는 것을 기억하기 바란다.

In [0]:
x_whole = x_front + x_middle + x_rear
print(x_whole)
x_both = x_even + x_odd
print(x_both)

[0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
[0.1, 0.3, 0.5, 0.7, 0.9, 0, 0.2, 0.4, 0.6, 0.8, 1.0]


리스트에 `*` 연산이 가능한데 반드시 정수를 곱해야 한다.
리스트에 정수를 곱하면 동일한 리스트 정수개가 `+` 연산된 긴 리스트가 만들어진다.
정수를 곱하지 않으면 `TypeError` 오류가 발생한다.

In [0]:
x_front_twice = x_front * 2  # x_front + x_front
print(x_front_twice)

[0, 0.1, 0.2, 0.3, 0.4, 0, 0.1, 0.2, 0.3, 0.4]


리스트의 * 연산은 다음과 같이 동일한 값을 반복 입력할 때 유용하다.

In [0]:
parks = ['park'] * 10
print(parks)

['park', 'park', 'park', 'park', 'park', 'park', 'park', 'park', 'park', 'park']


In [0]:
x_reat_trice = x_rear * 3.4

TypeError: can't multiply sequence by non-int of type 'float'

## `in` 연산자

`in` 연산자는 어떤 값이 리스트(또는 자료구조)에 저장된 값인지를 검사하고 그 결과를 논리값으로 준다.
`in` 연산자가 사용된 표현식은 조건문과 반복문에서 자주 사용된다.

In [0]:
if (0.9 in x_whole):   # 값이 집합의 원소인지를 검사하는 표현식
    msg = '집합의 원소이다'
else:
    msg = '집합의 원소가 아니다'
print(msg)
ans = 5.2 in x_front
print(ans)

집합의 원소이다
False


### 연습문제

1. 친구 10명의 이름이 저장된 리스트 friends를 만드시오.

2. 새로운 친구의 이름 "박중양"을 friends에 추가하시오.

3. 리스트에 단순한 값이 아닌 리스트를 저장할 수 있다.
friends에 친구 이름만 저장하지 않고 각 친구의 이름과 전화번호가 저장된 리스트를 저장하시오.
그러니까 friends를 리스트의 리스트로 만드는 것이다.
일반적으로 자료구조에 자료구조를 저장할 수 있다.

4. friends에 저장된 세 번째 친구의 전화번호는 무엇인가?

5. friends에 저장된 다섯 번째 친구의 이름은 무엇인가?

## `range()` 내장 함수

초항이 정수이고 공차가 정수인 등차 수열이 저장된 자료구조는 `range()` 함수를 사용하면 손쉽게 만들 수 있다.
`range()` 함수는 **범위**(range) 자료구조를 만들어 주는데 범위 자료구조가 리스트는 아니지만 리스트와 동일한 방법으로 사용할 수 있다.
필요하면 `list()` 함수를 이용해서 범위 자료구조를 리스트 자료구조로 바꿀 수 있다.

`range()`를 호출할 때는 `range(start, stop, step)`과 같이 초항, 최대값, 공차를 순서대로 정수로 주어야 한다.

어떤 함수를 실행할 때 괄호 속에 주는 값을 **인자**(argument)라고 한다.
앞에서 사용한 `print()`, `int()`, `float()`, `bool()` 함수 등을 실행할 때도 적절한 인자를 주었다는 것을 기억하기 바란다.
함수에 따라 인자의 개수와 자료형이 다르며, 이는 함수를 작성할 때 결정된다.

일반적으로 **함수를 호출(실행)하면 함수는 인자를 받아서 어떤 일을 한 다음 그 결과를 호출한 곳에 돌려준다.**
그러면 **함수를 호출한 코드는 함수가 돌려준 결과로 대체**되어 실행이 계속된다.
예를 들어 앞에서 배운 `int()` 함수는 인자를 정수로 변환한 결과를 돌려주고, `float()` 함수는 인자를 실수로 변환한 결과를 반환하며, `bool()` 함수는 인자를 논리값으로 변환한 결과를 돌려준다.
프로그래머는 **함수가 돌려주는 값을 변수에 저장해서 다른 일에 사용하는 것이 일반적**이다.

이에 반해 `print()` 함수는 인자를 모니터에 출력하는 일을 하지만 아무런 결과도 돌려주지 않는다.

**함수를 호출할 때 어떤 인자를 주어야 하는지 그리고 함수가 어떤 결과를 돌려주는지는 함수를 작성할 때 결정된다.**
지금까지 배운 내장 함수들은 이미 작성된 함수들이므로 어떤 인자를 어떻게 주어야 하며 어떤 값을 돌려주는지는 이미 정해져 있다.

In [0]:
x = float('123.45')
ret = print(x)
print(ret)

123.45
None


Note: `None`은 **결측치(missing value)**를 나타내는 파이썬 상수이다. `math`에 정의된 상수 `nan`도 결측치를 나타내는 상수이다.

`range()` 함수는 여러 가지 방법으로 실행할 수 있다.
초항이 0, 공차가 1이면 stop 인자만으로 `range(stop)`와 같이 실행할 수 있고, 공차가 1이면 start, stop 두 개의 인자로 `range(start, stop)`와 같이 실행할 수 있다.
이는 초항과 공차를 인자로 주지 않으면 자동적으로 0과 1로 설정된다는 의미이다.
즉, `range(stop)`은 `range(0, stop, 1)`과 같다.

**`range()` 함수로 생성한 범위 자료구조의 일부분을 추출하기 위해 인덱스와 슬라이스를 사용할 수 있다.**
원한다면 `list()` 함수로 범위 자료구조를 리스트 자료구조로 변환할 수도 있다.
리스트 자료구조는 다시 `set()`, `tuple()` 함수로 집합, 순서쌍으로 변환할 수 있다.
리스트, 집합, 순서쌍 자료구조는 서로 변환할 수 있다.

In [0]:
die = range(1, 7, 1)
print(type(die))
print(die)

<class 'range'>
range(1, 7)


범위 자료구조에 슬라이싱을 한 결과도 범위 자료구조이다.

In [0]:
print(die[3])
print(die[2:])

4
range(3, 7)


`list()` 함수를 이용하면 범위 자료구조를 리스트로 변환할 수 있다.

In [0]:
die = list(die)
print(type(die))
print(die)
print(len(die))

<class 'list'>
[1, 2, 3, 4, 5, 6]
6


In [0]:
lotto = list(range(1, 46))
print(lotto)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45]


범위 자료구조에 저장된 값의 개수를 구할 때도 `len()` 함수를 사용한다.

### 연습문제

1. 3, 6, 9, ..., 99가 저장된 리스트를 만드시오.

2. 2, 6, 10, 14, ..., < 1000이 저장된 리스트를 만드시오.

3. 앞 연습문제에서 만든 리스트를 집합, 순서쌍 자료구조로 변환하고 그 자료형을 확인하시오.


## 반복문

$\theta = 2$일 때 $0 \le x \le 10$ 구간에서 함수 $f(x) = \theta e^{-\theta x}$의 그래프를 작성하는데 필요한 $0 \le x \le 10$ 구간을 100 등분하는 값은 다음과 같이 구할 수 있다.

1. `range()` 함수로 $0, 1, \dots, 100$이 저장된 리스트를 만든다.
2. 리스트에 저장된 각 값을 10으로 나눈 결과가 저장된 리스트를 만든다.

In [0]:
seq = list(range(101))
print(seq[:10], seq[(len(seq)-10):])

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9] [91, 92, 93, 94, 95, 96, 97, 98, 99, 100]


리스트에 저장된 각 값을 10으로 나눈 결과가 저장된 리스트는 다음과 같이 구할 수 있다.

1. 빈 리스트를 새로 만든다.
2. 리스트에 저장된 값을 **순차적**으로 10으로 나눈 결과를 새로 만든 리스트에 추가한다.

리스트에 저장된 임의의 값을 변수로 처리하면 이 값을 10으로 나누는 표현식은 변하지 않는다.
동일한 연산을 리스트에 저장된 값만 순환하면서 반복하면 되는 것이다.
이와 같이 동일한 일을 반복할 때 사용하는 문장이 반복문이다.

파이썬의 대표적인 반복문은 다음과 같이 사용하는 `for` 문이다.
여기서 elem는 리스트에 저장된 값 하나를 가리키는 변수이며 변수 이름은 자유롭게 할 수 있다.

```python
for elem in 리스트:
    elem를 사용하는 한 줄 이상의 들여쓴 코드
```

파이썬의 자료구조는 객체지향적인 방법으로 구현되어 있다.
이는 자료구조가 자신이 저장하고 있는 자료를 처리할 수 있는 기능을 가지고 있음을 의미한다.
프로그래밍에서 기능은 함수로 구현하는데, 자료구조와 같은 객체가 가지고 있는 함수를 **메소드**(method)라고 한다.
객체의 메소드는 **`객체이름.메소드()`**와 같은 방법으로 호출한다.

리스트 자료구조는 자신의 끝에 자료를 하나 추가하는 메소드 `append()`를 가지고 있다.
이 메소드는 리스트 끝에 추가할 자료를 인자로 호출한다.
 `append()` 메소드를 이용하면 다음과 같이 프로그램 할 수 있다.

In [0]:
x = []
for i in seq:
    x.append(i/10) 

print(x)
print(len(x))
print(x[:5])
    

[0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8, 7.9, 8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9.0, 9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9, 10.0]
101
[0.0, 0.1, 0.2, 0.3, 0.4]


리스트 간의 덧셈 연산을 이용하면 다음과 같이 프로그램 할 수 있다.

In [0]:
x = []
for number in seq:
    value = number / 10
    x = x + [value]
print(x)

[0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8, 7.9, 8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9.0, 9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9, 10.0]


다음과 같이 더 간단하게 프로그램 할 수도 있다.

In [0]:
x = []
for number in seq:
    x = x + [number / 10]
print(x)

[0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8, 7.9, 8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9.0, 9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9, 10.0]


다음과 같이 인덱스를 이용해서 프로그램 할 수도 있다.

In [0]:
x = []
for idx in range(len(seq)):
    value = seq[idx] / 10
    x = x + [value]  # x.append(value)
print(x)

[0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8, 7.9, 8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9.0, 9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9, 10.0]


In [0]:
x = []
for idx in range(len(seq)):
    x = x + [seq[idx] / 10]
print(x)

[0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8, 7.9, 8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9.0, 9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9, 10.0]


위에 나열한 여러 가지 프로그램 중에서 어떤 것이 좋은가 하는 문제를 생각해봐야 한다.
이에 대해서는 다음에 논의하기로 하겠다.

**리스트에 저장된 값과 인덱스가 동시에 필요할 때는 `enumerate()` 내장함수를 사용한다.**
`enumerate()` 내장함수에 리스트를 인자로 주면 인덱스와 값을 따로 받아서 프로그램 할 수 있다.
`enumerate()` 함수의 사용법은 다음과 같다.

```python
for index, value in enumerate(리스트):
    index와 value를 이용한 한 줄 이상의 들여쓴 코드
```

$0 \le x \le 10$ 구간을 100등분한 값 중에서 9 이상인 값과 그 값의 인덱스는 다음과 같이 출력할수 있다.

In [0]:
for idx, x0 in enumerate(x):
    if x0 >= 9:
        print(idx, x0)

90 9.0
91 9.1
92 9.2
93 9.3
94 9.4
95 9.5
96 9.6
97 9.7
98 9.8
99 9.9
100 10.0


함수 값을 계산할 x값이 만들어졌으므로 함수 값은 다음과 같이 계산할 수 있다.

1. 함수 값을 저장할 빈 리스트 fx를 만든다.
2. 리스트 x에 저장된 값을 순회하면서 함수 값을 구해 fx에 추가한다.

함수의 값을 계산하는 표현식은 동일하므로 위의 두 단계에 걸쳐 함수의 값을 계산할 때도 반복문을 사용할 수 있다.

In [0]:
import math
theta = 2.

fx = []
for x0 in x:
    fx0 = theta * math.exp(-theta * x0)
    fx.append(fx0)

print(fx)

[2.0, 1.6374615061559636, 1.3406400920712787, 1.0976232721880528, 0.8986579282344431, 0.7357588823428847, 0.6023884238244043, 0.493193927883213, 0.40379303598931077, 0.33059777644317306, 0.2706705664732254, 0.22160631672466774, 0.18143590657882502, 0.14854715642866775, 0.12162012525043595, 0.09957413673572789, 0.08152440795673242, 0.06674653992065216, 0.05464744489458512, 0.0447415437123312, 0.03663127777746836, 0.029991153640955406, 0.024554679806136872, 0.020103671489267172, 0.01645949409804006, 0.013475893998170934, 0.011033128841521543, 0.009033161885225332, 0.007395727432965864, 0.006055109490751631, 0.004957504353332717, 0.004058861272591468, 0.003323114546347868, 0.0027207360750957877, 0.0022275502956896065, 0.0018237639311090325, 0.0014931716167533584, 0.0012225055222591446, 0.0010009028668812217, 0.0008194699579595736, 0.0006709252558050237, 0.0005493071399442851, 0.0004497346483576964, 0.0003682115873351584, 0.000301466150190953, 0.0002468196081733591, 0.00020207880367418684,

In [0]:
import math
theta = 2.0
fx = []
for x0 in x:
    f = theta * math.exp(-theta * x0)
    fx = fx + [f]
print(fx[:10])

[2.0, 1.6374615061559636, 1.3406400920712787, 1.097623272188053, 0.8986579282344431, 0.7357588823428847, 0.6023884238244043, 0.493193927883213, 0.40379303598931077, 0.33059777644317306]


## 또 다른 반복문: `while` 문

특정한 조건이 만족되는 동안(결과가 논리값인 표현식의 결과가 True이면)만 어떤 일을 반복하고 싶을 때는 `while` 문이 `for` 문보다 편리하다.

$\theta = 2$일 때 $0 \le x \le 10$ 구간에서 함수 $f(x) = \theta e^{-\theta x}$의 그래프를 작성하는 문제에서 $0 \le x \le 10$ 구간을 100 등분하는 101개의 값에서 함수 값을 계산하기로 했다.
$0 \le x \le 10$ 구간을 100 등분하는 101개의 값은 다음과 같이 구할 수 있다.

1. 빈 리스트 x를 만든다
2. 변수 x0에 0을 저장한다.
3. x0가 10 보다 작으면 x0를 리스트 x에 저장한다. x0가 10보다 크면 반복을 종료한다.
4. x0를 0.1만큼 증가시킨 다음 3으로 간다.

In [1]:
import math

lb_x = 0
ub_x = 10
n = 100
dx = (ub_x - lb_x) / n
theta = 2.0

x = []
fx = []

x0 = lb_x
while x0 <= ub_x:
    x.append(x0)
    fx0 = theta * math.exp(-theta*x0)
    fx.append(fx0)
    x0 = x0 + dx
    
print(x[:10])
print(fx[:10])


[0, 0.1, 0.2, 0.30000000000000004, 0.4, 0.5, 0.6, 0.7, 0.7999999999999999, 0.8999999999999999]
[2.0, 1.6374615061559636, 1.3406400920712787, 1.0976232721880528, 0.8986579282344431, 0.7357588823428847, 0.6023884238244043, 0.493193927883213, 0.4037930359893109, 0.3305977764431731]


반복문 내에서 특정한 조건이 만족되면 반복을 중단할 수 있다.
이때 `break` 문을 사용한다.

In [3]:
x = []
fx = []

x0 = lb_x
while True:
    if x0 <= ub_x:
        x.append(x0)
        fx0 = theta * math.exp(-theta*x0)
        fx.append(fx0)
    else:
        break
    x0 = x0 + dx
    
print(x[:10])
print(fx[:10])

[0, 0.1, 0.2, 0.30000000000000004, 0.4, 0.5, 0.6, 0.7, 0.7999999999999999, 0.8999999999999999]
[2.0, 1.6374615061559636, 1.3406400920712787, 1.0976232721880528, 0.8986579282344431, 0.7357588823428847, 0.6023884238244043, 0.493193927883213, 0.4037930359893109, 0.3305977764431731]


## 연습문제

1. $\theta = 2$일 때 $0 \le x \le 20$ 구간을 1000 등분한 점에서 함수 $f(x) = 1 - e^{-\theta x}$의 값을 계산하여 출력하시오.

2. $n=20$, $p=0.3$일 때 $x=0, 1, \dots, n$에서 함수 $f(x) = \binom{n}{x} p^x (1-p)^{n-x}$의 값을 계산하여 출력하시오.

3. $\alpha = 1.2$, $\beta=2.5$일 때 $-10 \le x \le 10$ 구간을 2000 등분한 점에서 함수 $p(x) = \frac{1}{1 + e^{-\alpha - \beta x}}$의 값을 계산하여 출력하시오.

4. 앞 문제에서 계산한 $p(x)$를 이용하여 $logit(x) = \log \left( \frac{p(x)}{1-p(x)} \right)$를 계산하여 출력하시오.

5. 구간 $[-5, 5]$을 100 등분하는 점에서 함수 $f(x) = \left\{
     \begin{array}{ll}
       x & \text{if } x \gt 0 \\
       0 & \text{if} x \le 0
     \end{array}
   \right.$를 계산하시오.

6. $\alpha=2$일 때 구간 $[-5, 5]$을 100 등분하는 점에서 함수 $f(x) = \left\{
     \begin{array}{ll}
       x & \text{if } x \gt 0 \\
       \alpha (e^x - 1) & \text{if} x \le 0
     \end{array}
   \right.$를 계산하시오.

7. 5명의 친구 이름이 저장된 리스트를 만드시오.
키보드에서 이름을 입력받고 이 이름이 친구 이름인지 아닌지를 검사하시오.
입력받은 이름이 친구 이름이면 "반갑다 친구야!"를 출력하고 친구 이름이 아니면 "누구세요?"를 출력하시오.

8. 한 사이트에 등록된 사용자는 5명이다.
5명 사용자의 이름과 암호가 저장된 리스트 usernames, passwords를 만드시오.

    * 사용자의 이름과 암호를 키보드에서 입력받아 정상적인 사용자인지를 확인하는 프로그램을 작성하시오.
    * 새로운 사용자의 이름과 암호를 입력받아서 usernames, passwords에 추가하는 프로그램을 작성하시오.
    
9. 사용자 이름과 암호를 별도의 리스트로 만들지 않고 하나의 리스트로 만들고자 한다.

    * 각 사용자의  이름과 암호를  리스트로 만들고 이 리스트들을 리스트 users에 저장하시오.
    * 사용자 이름을 입력받아서 users에 있는 사용자인지 검사하시오.
    * 입력받은 사용자 이름이 users에 있으면 해당 사용자의 암호를 입력받아서 올바른 암호인지 검사하시오.
    * 새로운 사용자의 이름과 암호를 users에 추가하시오.
    
10. `random` 모듈의 난수발생함수를 이용하여 10 명의 키와 체중이 저장된 리스트 heights, weights를 생성한 다음 10 명의 bmi가 저장된 리스트 bmis를 만들고 출력하시오.

11. 앞에서 만든 리스트 heights, weights에서 키가 180cm 이상인 사람의 키와 체중만 추출하여 별도의 리스트에 저장하시오.

12. 리스트 bmis를 이용하여 bmi가 25 이상인 사람의 키와 체중을 추출하여 별도의 리스트에 저장하시오.